### ISSUE TITLE
WebSocket with MultiplayerAPI spam a lot of errors unlike ENet

### ISSUE THREAD

**Issue by TIMONz1535** [godotengine/godot#44134]
**Godot version:** 3.2.3.stable.official

**OS/device including version:** Windows 10

**Issue description:**
If you use a **WebSocket** with **MultiplayerAPI** you will come across the fact that it spam tons of errors at any convenient situation. Which cannot be caught and handled in any way.
I spent about a week to figure out what I was doing wrong. I ran into the fact that when the client who is sending **rset** leaves the server, the server writes an error, and I cannot influence this in any way.
I found one project, it works great with **ENet**, but if you replace it with a **WebSocket**, you get the same errors right away.

**Steps to reproduce:**
1) Look at **Main.gd** line 41 and line 55. You may choose **NetworkedMultiplayerENet** peer or **WebSocketServer/Client** peer.
2) Launch separate applications **Server** and **Client**.
3) On **ENet** all works fine.
4) On **WebSocket** you get infinity errors on **Client** when **Server** app closed.
`ERROR: Trying to call an RPC via a network peer which is not connected.
   At: core/io/multiplayer_api.cpp:639`
And **Server** get errors when **Client** closed. Errors stop after a while.
`ERROR: put_packet: Condition "!is_connected_to_host()" is true. Returned: FAILED
   At: modules/websocket/wsl_peer.cpp:238`

**Minimal reproduction project:** 
[realtime-network-sync-godot-master.zip](https://github.com/godotengine/godot/files/5647970/realtime-network-sync-godot-master.zip)

**Comment by Calinou** [comment]
As far as I know, the WebSocket server is not a drop-in replacement for ENet. Many features available in ENet aren't available in WebSockets due to limitations of the protocol.

**Comment by Faless** [comment]
@Calinou  WebSocket can be used with the multiplayer API.

The main limitation is that it does not support unrealiable mode, so you have to be more mindful about how much data you send to avoid clogging it and getting lag/disconnection.

>  `ERROR: Trying to call an RPC via a network peer which is not connected. At: core/io/multiplayer_api.cpp:639`

This means what it means, i.e. that your are calling `rpc` while the network peer is not connected.
So, I would suggest trying to check if you get connected at all, or you if get a disconnect after a while e.g. because you just send too much data. And of course, do not call `rpc` when the peer is disconnected.

In general, this is a very bad issue report.

**Comment by TIMONz1535** [comment]
> ERROR: Trying to call an RPC via a network peer which is not connected. At: core/io/multiplayer_api.cpp:639

Ok, this problem is clear and solvable. It is not clear why using the same **high-level** multiplayer code, **WebSocket** behave differently.

> ERROR: put_packet: Condition "!is_connected_to_host()" is true. Returned: FAILED At: modules/websocket/wsl_peer.cpp:238

This error is much more complicated. If you make a player who will send **one rset** in **_physics_process**, then when you close the Client, you get error on Server in any case, why?
The player who disconnected from the server cannot send a **rset**. How to explain it logically? Did the packet come after the peer disconnected? Then why is it a programmer's fault, why doesn't the **WebSocket** just ignore it?

```
func _physics_process(_delta):
  if is_network_master():
    if get_tree().network_peer.get_connection_status() == NetworkedMultiplayerPeer.CONNECTION_CONNECTED:
      rset("puppet_pos", position)
```
This check dont help.

```
func _ready():
       get_tree().set_auto_accept_quit(false)

func _notification(event):
	if event == MainLoop.NOTIFICATION_WM_QUIT_REQUEST:
		wsClient.disconnect_from_host(1000, "Quit")
		get_tree().network_peer = null
		get_tree().quit()
	if event == MainLoop.NOTIFICATION_WM_GO_BACK_REQUEST:
		wsClient.disconnect_from_host(1000, "Quit")
		get_tree().network_peer = null
		get_tree().quit()
```
It helps, but if you close the process it won't work. But why, what is wrong, how to influence it? This is wrong behavior of **HIGH**-level network.

>In general, this is a very bad issue report.
Yeah bad. Surprisingly, almost all demos on the WebSocket work with low-level packets, and those that try to use the MultiplayerAPI have the same errors ¯\_(ツ)_/¯

**Comment by Faless** [comment]
There is an official multiplayer demo using websocket here:

https://github.com/godotengine/godot-demo-projects/tree/master/networking/websocket_multiplayer

Looking at your code, you keep making `rpc`s and `rset`s from the client after you get a disconnect, so you get the errors on the client.

The error on the server is trickier, I see how it could happen on broken pipes since it might get a disconnect on the first put_packet but the event will only be fired on the next poll. And yeah, the server shouldn't spam that way.
A proper fix is not straight forward though, maybe we can look at suppressing them in the `3.x` branch in the meantime.

**Comment by Faless** [comment]
This is fixed in current `master` via #66594 (WebSocket refactor).
